pkg/registry/common/expire/ns_server.go

```diff
@@ -30,88 +29,135 @@ import (
+	for event := range eventCh {
+		nse := event
+		if nse.ExpirationTime == nil {
+			continue
+		}
+
+		for i := 0; i < len(nse.NetworkServiceNames); i++ {
 			ns := nse.NetworkServiceNames[i]
-			var value int32
-			stored, _ := n.nsCounts.LoadOrStore(ns, &value)
-			if nse.ExpirationTime != nil && nse.ExpirationTime.Seconds < 0 {
-				atomic.AddInt32(stored, -1)
-			} else {
-				atomic.AddInt32(stored, 1)
+			state, ok := n.nsStates.Load(ns)
+			if !ok {
+				return
 			}
-			duration := time.Until(nse.ExpirationTime.AsTime())
-			timer := time.AfterFunc(duration, func() {
-				if atomic.AddInt32(stored, -1) <= 0 {
-					if ctx, ok := n.contexts.Load(ns); ok {
+			state.Lock()
+			timer, ok := state.Timers[nse.Name]
+			expirationDuration := time.Until(nse.ExpirationTime.AsTime().Local())
+			if !ok {
+				if expirationDuration > 0 {
+					state.Timers[nse.Name] = time.AfterFunc(expirationDuration, func() {
+						state.Lock()
+						ctx := state.Context
+						delete(state.Timers, nse.Name)
+						state.Unlock()
 						_, _ = n.Unregister(ctx, &registry.NetworkService{Name: ns})
+					})
+					state.Unlock()
+				}
+				continue
+			}
+
+			if expirationDuration < 0 {
+				delete(state.Timers, nse.Name)
+				if timer.Stop() {
+					if len(state.Timers) == 0 {
+						ctx := state.Context
+						state.Unlock()
+						_, _ = n.Unregister(ctx, &registry.NetworkService{Name: ns})
+						return
 					}
 				}
-			})
-			if v, loaded := n.timers.LoadOrStore(nse.Name, timer); loaded {
-				timer.Stop()
-				v.Stop()
-				v.Reset(duration)
+				state.Unlock()
+				continue
 			}
+
+			timer.Stop()
+			timer.Reset(expirationDuration)
+			state.Unlock()
 		}
 	}
 }
 
-func (n *nsServer) Register(ctx context.Context, request *registry.NetworkService) (*registry.NetworkService, error) {
+func (n *expireNSServer) Register(ctx context.Context, request *registry.NetworkService) (*registry.NetworkService, error) {
 	n.once.Do(func() {
-		go func() {
-			for n.monitorErr == nil && n.chainCtx.Err() == nil {
-				n.checkUpdates()
-			}
-		}()
+		c, err := n.nseClient.Find(n.chainCtx, &registry.NetworkServiceEndpointQuery{
+			NetworkServiceEndpoint: &registry.NetworkServiceEndpoint{},
+			Watch:                  true,
+		})
+		if err != nil {
+			n.monitorErr = err
+			return
+		}
+		go n.checkUpdates(registry.ReadNetworkServiceEndpointChannel(c))
 	})
+
 	if n.monitorErr != nil {
 		return nil, n.monitorErr
 	}
-	n.contexts.Store(request.Name, extend.WithValuesFromContext(n.chainCtx, ctx))
-	return next.NetworkServiceRegistryServer(ctx).Register(ctx, request)
+
+	resp, err := next.NetworkServiceRegistryServer(ctx).Register(ctx, request)
+
+	if err != nil {
+		return nil, err
+	}
+
+	valuesCtx := extend.WithValuesFromContext(n.chainCtx, ctx)
+
+	v, _ := n.nsStates.LoadOrStore(request.Name, &nsState{
+		Timers: make(map[string]*time.Timer),
+	})
+
+	v.Lock()
+	v.Context = valuesCtx
```

Comment:
Do we need to overwrite context on every new NS registered?

Reply:
I think yes, the chain can be changed.